omegaconf/grammar_visitor.py

```diff
@@ -0,0 +1,352 @@
+import sys
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    Generator,
+    Iterable,
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
+
+from antlr4 import TerminalNode
+
+from .errors import GrammarTypeError
+
+if TYPE_CHECKING:
+    from .base import Container, Node  # noqa F401
```

Comment:
I'd feel more comfortable if the visitor, which I consider very low level - did not call the Container/Node which are higher level - directly.

Can you switch to a callback mechanism?
This would also eliminate the need to pass in the resolve_args.

Send me a message if you are not sure what I mean by this.

Reply:
For `Container` I think I see what you mean -- essentially passing functions to the `GrammarVisitor` constructor that will be called to resolve interpolations.

For `Node`, it seems a bit trickier. I'm actually using it only for typing. This is because when we resolve an interpolation, the result may be a `Node` (ex: `${foo.bar}` returns a `Node` object), and we want to preserve it all the way up, so that for instance `${${x}.${y}}` returns the same node if `x` is set to `"foo"` and `y` is set to `"bar"`.

The easiest fix I can see to get rid of `Node` would be to loosen the typing and replace it with `Any` but it seems to me we would lose more than we gain by doing this. Thoughts?

Reply:
I switched to a callback mechanism in 16b96ed. You might want to check out this commit before continuing with the rest of the review, since it slightly changes some of the code that you haven't reviewed yet -- just to be aware of these changes when you get to it (or, alternatively, I can rebase on top of 76ebe1a -- the commit you said you reach -- and force push, if you prefer)